### PR TITLE
MRG: improve parallel backends

### LINKED ISSUE #210: MPIBackend does not allow passing whole objects
Currently, the MPIBackend passes the `params` dictionary to the child for running simulations. I think we should allow passing arbitrary Python objects as long as they are picklable. This will make it more similar to the Joblibs backend and also not limit code design. @blakecaldwell do you have an intuition whether this is easy/difficult to implement.

[Discovered this](https://github.com/jonescompneurolab/hnn-core/pull/199#discussion_r528268228) in #199 with @ntolley

### PR DESCRIPTION
**Pull request by jasmainak**
closes #210 

<s>This is mostly a PR stating my intent so far</s>

cc @blakecaldwell @ntolley @cjayb 

will ping again when ready, also attempting to improve the "bus factor" by messing with the code (not that we expect anything to happen to @blakecaldwell :)

### CODE REVIEW

File: hnn_core/parallel_backends.py [hunk 1]
```diff
@@ -337,10 +337,6 @@ def simulate(self, net):
             The Dipole results from each simulation trial
         """
 
-        # just use the joblib backend for a single core
-        if self.n_procs == 1:
-            return JoblibBackend(n_jobs=1).simulate(net)
```
Comment (jasmainak) [inline hnn_core/parallel_backends.py:342]: not quite sure why this was necessary?

Comment (blakecaldwell) [reply hnn_core/parallel_backends.py:342]: JoblibBackend for 1 job is essentially no backend. Don't bother creating the mpi child process for a single processor.

File: doc/parallel.rst [hunk 2]
```diff
@@ -73,13 +73,32 @@ Verifies that MPI, NEURON, and Python are all working together.
...
+   transfer has completed and it will verify the length of data it receives, printing a ``UserWarning`` if the
+   lengths don't match.
+
+It is important that ``MPISimulation`` uses the ``flush()`` method after each signal to ensure that the signal
+will immediately be available for reading by ``MPIBackend`` and not buffered with other output.
+
+Tests for parallel backends utilize a special ``@pytest.mark.incremental`` decorator (defined in ``conftest.py``)
+that causes a test failure to skip subsequent tests in the incremental block. For example, if a test running a
+simple MPI simulation fails, subsequent tests that compare simulation output between different backends will be
+skipped. These types of failures will be marked as a failure in CI.
```
Comment (jasmainak) [inline doc/parallel.rst:104]: just broke down the lines because it makes it easier to read.

**Comment by jasmainak** [comment]
ready for review. @ntolley now we pass the `Network` object instead of `net.params` so both backends are equivalent

File: hnn_core/parallel_backends.py [hunk 3]
```diff
@@ -69,6 +81,220 @@ def _read_all_bytes(fd, chunk_size=4096):
...
+            else:
+                total_time += 1
+                sleep(1)
+
+        # wait for an event on the selector, timeout after 1s
+        events = sel.select(timeout=1)
+        for key, mask in events:
+            callback = key.data
+            callback_result = callback(key.fileobj, mask)
+            if callback.__name__ == '_read_stdout':
```
Comment (jasmainak) [inline hnn_core/parallel_backends.py:227]: explicitly using the name of the functions to make the code a bit more transparent

File: hnn_core/parallel_backends.py [hunk 4]
```diff
@@ -69,6 +81,220 @@ def _read_all_bytes(fd, chunk_size=4096):
...
+            else:
+                raise ValueError("Completion signal from child MPI process"
+                                 " did not contain data length.")
+
+            return proc_data_bytes, data_len
+
+    return data
+
+
+def run_subprocess(command, pickled_obj, timeout, *args, **kwargs):
```
Comment (jasmainak) [inline hnn_core/parallel_backends.py:159]: tried to make something more generic. Didn't completely succeed, will try more in future PRs

File: hnn_core/parallel_backends.py [hunk 5]
```diff
@@ -345,105 +496,14 @@ def simulate(self, net):
...
-        # if simulation failed, raise exception
-        if proc.returncode != 0:
-            raise RuntimeError("MPI simulation failed")
-
-        sim_data = self._process_child_data(self.proc_data_bytes, data_len)
+        proc, proc_data_bytes, data_len = run_subprocess(
+            command=self.mpi_cmd, pickled_obj=pickled_net, timeout=4,
+            env=env, cwd=os.getcwd(), universal_newlines=True)
+
+        sim_data = _process_child_data(proc_data_bytes, data_len)
```
Comment (jasmainak) [inline hnn_core/parallel_backends.py:506]: one of the main objectives of this redesign was to have as little code in `MPIBackend.simulate()` as in `JoblibBackend.simulate()`

**Comment by jasmainak** [comment]
anyone got a chance to look? @blakecaldwell perhaps you're the best person to review this PR?

File: hnn_core/parallel_backends.py [hunk 6]
```diff
@@ -58,6 +58,18 @@ def _gather_trial_data(sim_data, net, n_trials):
...
 
 
+def _get_mpi_env():
+    """Set some MPI environment variables."""
+    my_env = os.environ.copy()
+    if 'win' not in sys.platform:
+        my_env["OMPI_MCA_btl_base_warn_component_unused"] = '0'
+
+    if 'darwin' in sys.platform:
+        my_env["PMIX_MCA_gds"] = "^ds12"  # open-mpi/ompi/issues/7516
```
Comment (ntolley) [inline hnn_core/parallel_backends.py:68]: Looks like these issues are dependent on the ompi version. Are there other issues that would necessitate **ompi>=ver** in the install instructions?

File: hnn_core/parallel_backends.py [hunk 7]
```diff
@@ -162,14 +390,10 @@ class MPIBackend(object):
         can be less than the user specified value if limited by the cores on
         the system, the number of cores allowed by the job scheduler, or
         if mpi4py could not be loaded.
-    mpi_cmd_str : str
-        The string of the mpi command with number of procs and options
-    proc_data_bytes: bytes object
-        This will contain data received from the MPI child process via stderr.
-
+    mpi_cmd : list
```
Comment (ntolley) [inline hnn_core/parallel_backends.py:393]: list of str?

File: hnn_core/parallel_backends.py [hunk 8]
```diff
@@ -345,105 +498,11 @@ def simulate(self, net):
...
-        os.close(pipe_stderr_r)
-        os.close(pipe_stderr_w)
-
-        # if simulation failed, raise exception
-        if proc.returncode != 0:
-            raise RuntimeError("MPI simulation failed")
-
-        sim_data = self._process_child_data(self.proc_data_bytes, data_len)
+        proc, sim_data = run_subprocess(
+            command=self.mpi_cmd, obj=net, timeout=4,
```
Comment (ntolley) [inline hnn_core/parallel_backends.py:504]: Curious where this timeout value came from. Clearly it works currently, but I wonder if this can fail if we add more `CellResponse` recordings in the future.

File: hnn_core/mpi_child.py [hunk 9]
```diff
@@ -54,20 +54,20 @@ def __exit__(self, type, value, traceback):
             from mpi4py import MPI
             MPI.Finalize()
 
-    def _read_params(self):
-        """Read params broadcasted to all ranks on stdin"""
+    def _read_obj(self):
```
Comment (ntolley) [inline hnn_core/mpi_child.py:57]: Is the goal here to disentangle HNN vocabulary from generic MPI processes? I don't see an immediate use for this outside of passing the `Network` around. Not a big issue though considering the calls to this function are pretty transparent.

**Review by ntolley — APPROVED** [review]
Just to make sure I understand, the major changes were: 
1) moving the member functions outside of the backends class
2) Passing the `Network` to the child processes

The changes seem good to me but I'm certainly no authority on parallelization. @blakecaldwell may have some insight if there are potential issues that could crop up with passing larger more complex objects to processes. 

I'm happy to merge if you're satisfied, unless you want to hold out on a review from the expert ;)

**Comment by blakecaldwell** [comment]
Hi all. Thanks @jasmainak for the effort here. I'm happy to review if it can wait about 1 week from now.

**Comment by jasmainak** [comment]
sure! no rush on this PR :)

File: hnn_core/mpi_child.py (hunk 9 above)
Comment (jasmainak) [reply hnn_core/mpi_child.py:57]: yes exactly ... if we can make things a bit more general/reusable, we can document and test them independently of the simulations. I don't think I achieved that though :)

File: hnn_core/parallel_backends.py (hunk 6 above)
Comment (blakecaldwell) [reply hnn_core/parallel_backends.py:68]: I don't foresee any. We don't have users install openmpi directly, so are somewhat dependent on the versions that work best for each OS (installed by conda or Microsoft MPI).

File: doc/parallel.rst [hunk 10]
```diff
@@ -73,13 +73,32 @@ Verifies that MPI, NEURON, and Python are all working together.
...
-
-Tests for parallel backends utilize a special ``@pytest.mark.incremental`` decorator (defined in ``conftest.py``) that causes a test failure to skip subsequent tests in the incremental block. For example, if a test running a simple MPI simulation fails, subsequent tests that compare simulation output between different backends will be skipped. These types of failures will be marked as a failure in CI.
\ No newline at end of file
+MPI parallelization with NEURON requires that the simulation be launched with the ``nrniv`` binary from the command-line.
+The ``mpiexec`` command is used to launch multiple ``nrniv`` processes which communicate via MPI. This is done using
+``subprocess.Popen()`` in ``MPIBackend.simulate()`` to launch parallel child processes (``MPISimulation``) to carry out
+the simulation. The communication sequence between ``MPIBackend`` and ``MPISimulation`` is outlined below.
+
+#. In order to pass the parameters from ``MPIBackend`` the child ``MPISimulation`` processes' ``stdin`` is used. 
+   Parameters are pickled and base64 encoded before being written to the processes' ``stdin``.
```
Comment (blakecaldwell) [inline doc/parallel.rst:82]: Update to mention new vocabulary about objects being passed, though still technically accurate for this PR

File: hnn_core/parallel_backends.py (hunk 8 above)
Comment (blakecaldwell) [reply hnn_core/parallel_backends.py:504]: It's not related to how much data is being passed. That is handled by the "end_of_data" signal. This just waits long enough to catch error messages if the process still has some in its stderr buffer. Also, it's unlikely that there are any errors at this point, during the shutdown process, but this just covers that case.

**Review by blakecaldwell — APPROVED** [review]
Looks good to me. It's hard to follow the changes made to handling the callback return values because "_read_stderr" and "_read_stdout" were moved outside of the classes, but I didn't see functional differences there.

I think it's a solid improvement @jasmainak!

**Comment by blakecaldwell** [comment]
@jasmainak what do you think is a path forward here? some of these changes are already part of master (e.g. passing `net` instead of `params`).

**Comment by jasmainak** [comment]
yeah, not sure if it's worth getting back to this PR. Specially since you're planning to change it to the `Queue` system. I'm fine closing this pull request.

**Comment by jasmainak** [comment]
closed by #296
